clang/include/clang/AST/StmtOpenACC.h

```diff
@@ -0,0 +1,142 @@
+//===- StmtOpenACC.h - Classes for OpenACC directives  ----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+/// \file
+/// This file defines OpenACC AST classes for statement-level contructs.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_AST_STMTOPENACC_H
+#define LLVM_CLANG_AST_STMTOPENACC_H
+
+#include "clang/AST/Stmt.h"
+#include "clang/Basic/OpenACCKinds.h"
+#include "clang/Basic/SourceLocation.h"
+
+namespace clang {
+/// This is the base class for an OpenACC statement-level construct, other
+/// construct types are expected to inherit from this.
+class OpenACCConstructStmt : public Stmt {
+  friend class ASTStmtWriter;
+  friend class ASTStmtReader;
+  /// The directive kind. Each implementation of this interface should handle
+  /// specific kinds.
+  OpenACCDirectiveKind Kind = OpenACCDirectiveKind::Invalid;
+  /// The location of the directive statement, from the '#' to the last token of
+  /// the directive.
+  SourceRange Range;
+
+  // TODO OPENACC: Clauses should probably be collected in this class.
+
+protected:
+  OpenACCConstructStmt(StmtClass SC, OpenACCDirectiveKind K,
+                       SourceLocation Start, SourceLocation End)
+      : Stmt(SC), Kind(K), Range(Start, End) {}
+
+public:
+  OpenACCDirectiveKind getDirectiveKind() const { return Kind; }
+
+  static bool classof(const Stmt *S) {
+    return S->getStmtClass() >= firstOpenACCConstructStmtConstant &&
+           S->getStmtClass() <= lastOpenACCConstructStmtConstant;
+  }
+
+  SourceLocation getBeginLoc() const { return Range.getBegin(); }
+  SourceLocation getEndLoc() const { return Range.getEnd(); }
+
+  child_range children() {
+    return child_range(child_iterator(), child_iterator());
+  }
+
+  const_child_range children() const {
+    return const_cast<OpenACCConstructStmt *>(this)->children();
+  }
+};
+
+/// This is a base class for any OpenACC statement-level constructs that have an
+/// associated statement. This class is not intended to be instantiated, but is
+/// a convenient place to hold the associated statement.
+class OpenACCAssociatedStmtConstruct : public OpenACCConstructStmt {
+  friend class ASTStmtWriter;
+  friend class ASTStmtReader;
+  template <typename Derived> friend class RecursiveASTVisitor;
+  Stmt *AssociatedStmt = nullptr;
```

Comment:
I hhave a question about children stmts/exprs here. Does OpenACC nodes may have other children rather than AssociatedStmt? If it is possible, better to use tail allocation and store AssociatedStmt as part of the tail-allocated memory space

Reply:
As far as I can tell, they do NOT have any.  Individual clauses have expressions (and some of the other nodes have expressions), but there is never more than 1 statement associated with it.

I considered the trailing storage, but since it is always 'required' and only on certain constructs  (which I'll select via giving them a different AST node), I think a member for the `ComputeConstruct`, and a few others makes the most sense?  But if you see a good reason to switch to trailing storage during future review, please let me know, and I'll switch it over!

Reply:
The problem here, that if you're going to implement it in some other level of abstraction, rather than here, you will need to migrate AssociatedStmt to the same level to make children() correctly return list of child stmt/exprs. I don't know what's your plan/design approach, so it is absolutely up to you for now. I'm just bringing this to make you aware of possible issues here.

Reply:
> The problem here, that if you're going to implement it in some other level of abstraction, rather than here, you will need to migrate AssociatedStmt to the same level to make children() correctly return list of child stmt/exprs. I don't know what's your plan/design approach, so it is absolutely up to you for now. I'm just bringing this to make you aware of possible issues here.

Thanks!  I'll keep it in mind... I presume there will be quite a few things that end up getting 'moved around' a bit as I get deeper into implementation, so this is good to know!

There are only two constructs with a directly associated expression (`cache` and `wait`) so those might need trailing storage (and there is `routine`, but that is an associated name which will turn into a decl), but neither of those have an associated stmt.